### DISCUSSION TITLE
OpenAPI schema is not valid

### DISCUSSION THREAD

**Discussion by LKay** [fastapi/fastapi#8188] [Questions]
**Describe the bug**

The `openapi.json` generated from the docs is not valid OpenAPI 3.0 schema and doesn't pass validation. Can be checked against few available validators:

- https://apidevtools.org/swagger-parser/online/
- https://editor.swagger.io/

**Expected behavior**

Passes validation.

**Environment:**
 - OS: macOS
 - FastAPI Version: 0.33.0
 - Python version: 3.7.3

**Additional context**

After looking at the schema I found some noticeable parts:

- Security definitions are duplicated multiple times is there are some dependencies with shared sub dependencies. In the schema it looks like:

```json
"security": [
  {
    "My Auth": []
  },
  {
    "My Auth": []
  }
]
``` 

- For numeric values min/max validation flags should be boolean values instead of integers:

```json
"schema": {
  "title": "Size",
  "maximum": 100,
  "exclusiveMinimum": 0,
  "type": "integer",
  "description": "Number of records to return",
  "default": 10
}
```
The `"exclusiveMinimum": 0,` should be in fact `"exclusiveMinimum": false,`

- Path parameters is referenced in multiple dependencies for a route get duplicated:

```json
"parameters": [
  {
    "required": true,
    "schema": {
      "title": "User_Id",
      "type": "string",
      "format": "uuid"
    },
    "name": "user_id",
    "in": "path"
  },
  {
    "required": true,
    "schema": {
      "title": "User_Id",
      "type": "string",
      "format": "uuid"
    },
    "name": "user_id",
    "in": "path"
  }
]
```

- References are not properly defined and thus resolved, resulting in multiple errors such as:

```
Missing required property: $ref at #/user/roles
```

**Comment by euri10** [comment]
Would you have a minimal example @LKay please, I just tried to validate openapi.json generated from 3 of my current projects and got no errors, that would be useful to advance on the topic.

**Comment by LKay** [comment]
Here is the minimal example that generates all but the `$ref` issues I mentioned above:

```python
from uuid import UUID

from authlib.jose import jwt, JWTClaims
from fastapi import APIRouter, Body, Depends, Security, FastAPI
from fastapi.security import OAuth2PasswordBearer, SecurityScopes
from pydantic import BaseModel, condecimal, Decimal


class AccessToken(JWTClaims):
    pass


class AuthData(BaseModel):
    jwt: str
    claims: AccessToken


oauth_scheme = OAuth2PasswordBearer(
    tokenUrl="https://auth.domain.com/token",
    scheme_name="My Auth",
    scopes={
        "email": "Email address access"
    }
)


async def require_oauth(
        scopes: SecurityScopes,
        bearer: str = Depends(oauth_scheme)
) -> AuthData:
    claims: AccessToken = jwt.decode(bearer, "key", claims_cls=AccessToken)
    claims.validate()

    return AuthData(
        jwt=bearer,
        claims=claims
    )


class User(BaseModel):
    pass


async def current_user(*,
                       auth: AuthData = Security(require_oauth)) -> User:
    # Get user from the database based on auth claims
    return User()


async def user_exists(*,
                      user_id: UUID) -> bool:
    # Query database for existence of requested user
    return True


class BugData(BaseModel):
    positive_decimal: condecimal(gt=Decimal(0))


router = APIRouter()


@router.post(
    "/users/{user_id}/bug",
    dependencies=[
        Depends(user_exists)
    ]
)
async def bug(*,
              user_id: UUID,
              user: User = Depends(current_user),
              data: BugData = Body(...)):
    pass


app: FastAPI = FastAPI(
    title="Bug API"
)

app.include_router(
    router,
    dependencies=[
        Security(require_oauth)
    ]
)
```

This is simplified for the sake of reproducing but in my actual app I have quite a few routes dependencies which depend on other dependencies, similar way as `current_user` in the example. The schema generated from the above does't seem to be valid.

**Comment by euri10** [comment]
you got user_id twice (once in the path `/users/{user_id}/bug` and once in the dependency) which I think may cause an issue in the schema : should you modify your dependency to 
```
async def user_exists(*,
                      another_user_id: UUID) -> bool:
    # Query database for existence of requested user
    return True
```
the validation error disappears, not sure how it may impact the rest of your code.

as for the other one, you should read https://github.com/tiangolo/fastapi/issues/240

**Comment by LKay** [comment]
In the above case both `user_id` are actually the same value coming originally from path, it’s just used twice in different places. Once to fetch some user inside dependency and second is passed to use inside a route logic. That said it should not duplicated it in schema definition. Renaming fixes the schema but it creates another redundant, required parameter on that route, which shouldn’t happen.

**Comment by euri10** [comment]
effectively my bad, wrong "fix" sorry

**Comment by tiangolo** [comment]
Thanks for the report @LKay ! I had to modify your example a bit to make it a minimum self-contained app, not requiring `authlib` to demonstrate the issue.

I wasn't able to replicate your auth errors, I imagine it was probably fixed recently.

About `exclusiveMinimum`, and `exclusiveMaximum`, those are 2 special cases, as in previous versions of JSON Schema they were booleans. But in recent versions those are integers.

OpenAPI is based on JSON Schema, but it was based on the initial version, before that change.

So, for those 2 specific keywords, we could support a recent version of JSON Schema or the older version used by OpenAPI. FastAPI (and actually Pydantic) use the recent version of JSON Schema. This allows, for example, using the integrated schemas with tools that can generate UIs based on JSON Schema. Still, `minimum` and `maximum` work the same in both specifications, so it's only those two specific keywords. You can track the progress about that in this issue: https://github.com/tiangolo/fastapi/issues/240

About the duplicated parameter ID in OpenAPI, that is indeed a bug :bug: . I created a new issue with a simple example to reproduce it and an included test to verify it: #967 

Also, thanks @euri10  for the help here!

I'm gonna close this issue now, let's keep track of that bug in the new issue.

**Comment by Kyle-sandeman-mrdfood** [comment]
Hi, this is still an issue 3 years after the last comment.

Here's a code snippet from the FastAPI docs:
```
from fastapi import FastAPI, Path
from typing_extensions import Annotated

app = FastAPI()


@app.get("/items/{item_id}")
async def read_items(
    item_id: Annotated[int, Path(title="The ID of the item to get", gt=0, le=1000)],
    q: str,
):
    results = {"item_id": item_id}
    if q:
        results.update({"q": q})
    return results
```

requirements.txt: (latest versions at time of writing)
```
fastapi==0.95.1
pydantic==1.10.7
```

and pasting the /openapi.json into https://apitools.dev/swagger-parser/online/ leaves you with multiple schema failures.
```
Swagger schema validation failed.
  #/paths/~1items~1{item_id}/get/parameters/0/schema/exclusiveMinimum must be boolean
  #/paths/~1items~1{item_id}/get/parameters/0/schema must have required property '$ref'
  #/paths/~1items~1{item_id}/get/parameters/0/schema must match exactly one schema in oneOf
  #/paths/~1items~1{item_id}/get/parameters/0 must have required property '$ref'
  #/paths/~1items~1{item_id}/get/parameters/0 must match exactly one schema in oneOf
```

From what I can see, fastapi/pydantic is setting the OpenAPI version to 3.0.2 which is not valid for the json schema version it is apparently using. OpenAPI 3.1 supports the numeric exclusiveMinimum. I don't know enough about the `$ref` part to comment.
If the decision to use a recent version of json schema is paramount to other functionality, the OpenAPI version should be bumped to a compatible one.
